admin/server/auth/pkce.go

```diff
@@ -0,0 +1,178 @@
+package auth
+
+import (
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/rilldata/rill/cli/pkg/auth"
+)
+
+const authorizationCodeGrantType = "authorization_code"
+
+// in-memory storage for authorization codes
+// TODO use persistent storage to handle admin server restarts, but since codes expire in a minute, is it really necessary, user can just re-initiate login ?
+var authCodeDB = make(map[string]AuthorizationCode)
```

Comment:
It's necessary because we have multiple replicas of the admin service – so the next request may go to another machine